### PR TITLE
Remove vault cache when reloading dataset

### LINKED ISSUE #6990: Reloading a dataset layer should cache-clear the corresponding data vaults
The vaultCache entries for a layer should be removed when the layer reload button is used.

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Reload layer, see log

### Issues:
- fixes #6990

------
(Please delete unneeded items, merge only when none are left open)
- [x] Needs datastore update after deployment

### CODE REVIEW

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala [hunk 1]
```diff
@@ -227,4 +232,22 @@ class DataSourceService @Inject()(
...
 
+  def invalidateVaultCache(dataSource: InboxDataSource, dataLayerName: String): Fox[Unit] =
+    for {
+      genericDataSource <- dataSource.toUsable
+      dataLayer <- genericDataSource.getDataLayer(dataLayerName)
+      magsOpt = dataLayer match {
+        case layer: N5Layer          => Some(layer.mags)
+        case layer: PrecomputedLayer => Some(layer.mags)
+        case _: SegmentationLayer    => None
+        case _: WKWLayer             => None
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala:243]: Is there a reason to have these here? Shouldn’t they match to the `_` below? Also, SegmentationLayers may be ZarrLayers at the same time, I think

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala [hunk 2]
```diff
@@ -227,4 +232,22 @@ class DataSourceService @Inject()(
...
+        case layer: ZarrLayer        => Some(layer.mags)
+        case _                       => None
+      }
+      _ = magsOpt match {
+        case Some(mags) => mags.map(mag => dataVaultService.clearVaultPath(mag))
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala:248]:
```suggestion
        case Some(mags) => mags.map(mag => dataVaultService.removeVaultFromCache(mag))
```

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala [hunk 3]
```diff
@@ -484,7 +484,7 @@ Expects:
           organizationName)
         for {
           _ <- Fox.runIf(layerName.isDefined)(
-            dataSourceRepository.invalidateVaultCache(organizationName, dataSetName, layerName.getOrElse("")))
+            dataSourceService.invalidateVaultCache(reloadedDataSource, layerName.getOrElse("")))
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala:487]:
```suggestion
            dataSourceService.invalidateVaultCache(reloadedDataSource, layerName))
```
Would it be possible to do all layers if no layer name is passed to the route?

**Review by fm3 — COMMENTED** [review]
Thanks for having a look at this! I think the structure is good already :) Please see my two comments. Also: could this return the number of removed keys? Then we could add that to the logging in the controller.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala (hunk 1 above)
Comment (frcroth) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DataSourceService.scala:243]: This was the automatic IntelliJ exhaustive match, will be adjusted.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala (hunk 3 above)
Comment (frcroth) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala:487]: > Thanks for having a look at this! I think the structure is good already :) Please see my two comments. Also: could this return the number of removed keys? Then we could add that to the logging in the controller.

For the number of keys? Do we need to check if it exists before removing or would it be fine to assume it exists, so counting every remove call as one removed vault. In that path, the number will just be the number of mags

Comment (fm3) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/controllers/DataSourceController.scala:487]: I guess the number of mags will do for the moment. I added a note to https://github.com/scalableminds/webknossos/issues/6969 that the cache API’s remove should return the number of removed items (0 or 1). If we add that there later we can make this more exact here

**Review by fm3 — APPROVED** [review]

**Pull request merged**